### PR TITLE
Logs Panel: Table UI - Reordering table columns via drag-and-drop

### LINKED ISSUE #78837: Logs Panel: Table UI - Reorder columns
Currently in the table UI, the fields in the multi-select sidebar are sorted alphabetically, but the actual order of the columns in the table appears to be set by their order in the labels object.
image

Give users a way to reorder the columns in the table visualization!

 - [ ] Users can reorder columns
 - [ ] Column ordering is respected when added to dashboard panel
 - [ ] Column ordering is respected when sharing an explore URL

### PR DESCRIPTION
**Pull request by gtk-grafana**
**What is this feature?**
Adding react-beautiful-dnd to field selection UI to facilitate ordering columns in table UI

**Why do we need this feature?**
The table UI facilities users to make associations between field values; customizing the order of fields is going to provide our users with the flexibility to do just that.

**Who is this feature for?**
Logs table users in explore

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/78837

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/explore/Logs/LogsTableMultiSelect.tsx [hunk 1]
```diff
@@ -57,13 +58,16 @@ export const LogsTableMultiSelect = (props: {
...
+          reorderColumn={props.reorderColumn}
+          id={'selected-fields'}
           toggleColumn={props.toggleColumn}
           labels={props.filteredColumnsWithMeta ?? props.columnsWithMeta}
           valueFilter={(value) => props.columnsWithMeta[value]?.active ?? false}
         />
 
         <div className={styles.columnHeader}>Fields</div>
         <LogsTableNavColumn
+          id={'available-fields'}
```
Comment (gtk-grafana) [inline public/app/features/explore/Logs/LogsTableMultiSelect.tsx:70]: not required anymore

File: public/app/features/explore/Logs/LogsTableWrap.tsx [hunk 2]
```diff
@@ -164,19 +177,28 @@ export function LogsTableWrap(props: Props) {
       labels.forEach((labels: Labels) => {
         const labelsArray = Object.keys(labels);
         // Iterate through the label values
-        labelsArray.forEach((label) => {
+        labelsArray.forEach((label, index) => {
```
Comment (gtk-grafana) [inline public/app/features/explore/Logs/LogsTableWrap.tsx:180]: index unused?

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import { fieldNameMeta } from './LogsTableWrap';
+
+export function getLogsFieldsStyles(theme: GrafanaTheme2) {
+  return {
+    wrap: css({
+      marginTop: theme.spacing(1),
+      marginBottom: theme.spacing(1),
+      display: 'flex',
+    }),
+    // Making the checkbox sticky and label scrollable for labels that are wider then the container
```
Comment (gtk-grafana) [inline public/app/features/explore/Logs/LogsTableActiveFields.tsx:19]: comments no longer relevant

**Comment by gtk-grafana** [comment]
Bug: looks like the position of the field is preventing the query filters from working in the table cell.
image
image

**Comment by gtk-grafana** [comment]
Was able to resolve the above bug by passing through the logsFrame before the transformations are applied

**Comment by gtk-grafana** [comment]
image

We added the drag icon used in the transformation UI to each selected field, and removed the percentage from active fields.

@niat22 Would love to review this when you have some time!

**Review by svennergr — COMMENTED** [review]
I think the color of the buttons does not match the color that is used at other places:
image
image
Do we want to change that?
---
I am not sure if this is an issue with this PR, but it looks like only labels from the first query are displayed?

https://github.com/grafana/grafana/assets/8092184/bdb66613-0cba-41bc-b0d2-1f496006fbd5

**Comment by svennergr** [comment]
> @svennergr did you see the select for picking which query is visualized? The table only displays one frame at a time right now. image

I know I missed something obvious, thanks - and sorry! Will do another review in a bit.

**Comment by gtk-grafana** [comment]
No worries! And good catch, it looks like the transformation editor sets the opacity of that element to `opacity: 0.4;`. I'll push that up!

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  if (labelKeys.length) {
+    // If we have a reorderColumn function, we need to wrap the nav items in dnd components
+    return (
+      <DragDropContext onDragEnd={onDragEnd}>
+        <Droppable droppableId={'order-fields'} direction="vertical">
+          {(provided) => (
+            <div className={styles.columnWrapper} {...provided.droppableProps} ref={provided.innerRef}>
+              {labelKeys.sort(sortLabels(labels)).map((labelName, index) => (
+                <Draggable draggableId={labelName} key={labelName} index={index}>
+                  {(provided: DraggableProvided) => (
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableActiveFields.tsx:71]: Would be great to get a slightly different background color - maybe just seconday? - on the draggable when it's being dragged. Looks like you could use something like this: https://github.com/atlassian/react-beautiful-dnd/blob/master/docs/api/draggable.md#2-snapshot-draggablestatesnapshot

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+  const { labels, valueFilter, toggleColumn } = props;
+  const theme = useTheme2();
+  const styles = getLogsFieldsStyles(theme);
+  const labelKeys = Object.keys(labels).filter((labelName) => valueFilter(labelName));
+  if (labelKeys.length) {
+    // If we have a reorderColumn function, we need to wrap the nav items in dnd components
+    return (
+      <DragDropContext onDragEnd={onDragEnd}>
+        <Droppable droppableId={'order-fields'} direction="vertical">
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableActiveFields.tsx:66]:
```suggestion
        <Droppable droppableId="order-fields" direction="vertical">
```

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            <div className={styles.columnWrapper} {...provided.droppableProps} ref={provided.innerRef}>
+              {labelKeys.sort(sortLabels(labels)).map((labelName, index) => (
+                <Draggable draggableId={labelName} key={labelName} index={index}>
+                  {(provided: DraggableProvided) => (
+                    <div
+                      className={styles.wrap}
+                      ref={provided.innerRef}
+                      {...provided.draggableProps}
+                      {...provided.dragHandleProps}
+                      title={`${labelName} appears in ${labels[labelName]?.percentOfLinesWithLabel}% of log lines`}
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableActiveFields.tsx:77]: If `labels[labelName]` is `undefined` we shouldn't display the title at all, right?

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+export const LogsTableActiveFields = (props: {
+  labels: Record<string, fieldNameMeta>;
+  valueFilter: (value: string) => boolean;
+  toggleColumn: (columnName: string) => void;
+  reorderColumn: (sourceIndex: number, destinationIndex: number) => void;
+  id: string;
+}): JSX.Element => {
+  const { reorderColumn } = props;
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableActiveFields.tsx:49]:
```suggestion
  const { reorderColumn, labels, valueFilter, toggleColumn } = props;
```

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+  const onDragEnd = (result: DropResult) => {
+    if (!result.destination || !reorderColumn) {
+      return;
+    }
+    reorderColumn(result.source.index, result.destination.index);
+  };
+
+  const { labels, valueFilter, toggleColumn } = props;
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableActiveFields.tsx:58]:
```suggestion
```

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  labels: Record<string, fieldNameMeta>;
+  valueFilter: (value: string) => boolean;
+  toggleColumn: (columnName: string) => void;
+  reorderColumn: (sourceIndex: number, destinationIndex: number) => void;
+  id: string;
+}): JSX.Element => {
+  const { reorderColumn } = props;
+
+  const onDragEnd = (result: DropResult) => {
+    if (!result.destination || !reorderColumn) {
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableActiveFields.tsx:52]:
```suggestion
    if (!result.destination) {
```
Shouldn't be needed, right?

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    }
+    reorderColumn(result.source.index, result.destination.index);
+  };
+
+  const { labels, valueFilter, toggleColumn } = props;
+  const theme = useTheme2();
+  const styles = getLogsFieldsStyles(theme);
+  const labelKeys = Object.keys(labels).filter((labelName) => valueFilter(labelName));
+  if (labelKeys.length) {
+    // If we have a reorderColumn function, we need to wrap the nav items in dnd components
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableActiveFields.tsx:63]: Is that comment still actual? `reorderColumn` seems to be mandatory now?

File: public/app/features/explore/Logs/LogsTableWrap.tsx [hunk 11]
```diff
@@ -35,11 +35,23 @@ interface Props extends Themeable2 {
   datasourceType?: string;
 }
 
-export type fieldNameMeta = {
+type activeFieldMeta = {
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableWrap.tsx:38]: I think based on our styleguide we should capitalise all of those type names at one point: https://github.com/grafana/grafana/blob/main/contribute/style-guides/frontend.md#types-and-interfaces 

Not necessary in this PR.

File: public/app/features/explore/Logs/LogsTableNavField.tsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      fontSize: theme.typography.pxToRem(11),
+      opacity: 0.6,
+    }),
+    contentWrap: css({
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'space-between',
+      width: '100%',
+    }),
+    // Making the checkbox sticky and label scrollable for labels that are wider then the container
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableNavField.tsx:31]: I think I couldn.t get the "scrollable" label to work. Did it work for you?
image

File: public/app/features/explore/Logs/LogsTableNavField.tsx [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      <div className={styles.contentWrap}>
+        <Checkbox
+          className={styles.checkboxLabel}
+          label={props.label}
+          onChange={props.onChange}
+          checked={props.labels[props.label]?.active ?? false}
+        />
+        {props.showCount && (
+          <button className={styles.labelCount} onClick={props.onChange}>
+            {props.labels[props.label]?.percentOfLinesWithLabel}%
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableNavField.tsx:65]: Same here, we shouldn't display this information at all, if `props.labels[props.label]` is undefined. I know this is probably quite an edge case, but who knows how people will use this component at one point 😆

File: public/app/features/explore/Logs/LogsTableWrap.tsx [hunk 14]
```diff
@@ -308,6 +363,74 @@ export function LogsTableWrap(props: Props) {
...
+  // Toggle a column on or off when the user interacts with an element in the multi-select sidebar
+  const toggleColumn = (columnName: fieldName) => {
+    if (!columnsWithMeta || !(columnName in columnsWithMeta)) {
+      console.warn('failed to get column', columnsWithMeta);
+      return;
+    }
+
+    const length = Object.keys(columnsWithMeta).filter((c) => columnsWithMeta[c].active).length;
+
+    const isActive = !columnsWithMeta[columnName]?.active ? true : undefined;
```
Comment (svennergr) [inline public/app/features/explore/Logs/LogsTableWrap.tsx:377]:
```suggestion
    const isActive = !columnsWithMeta[columnName].active ? true : undefined;
```
Shouldn't be needed

**Review by svennergr — COMMENTED** [review]
Left a few comments. Overall just some nits, feature works great! 🚀

File: public/app/features/explore/Logs/LogsTableActiveFields.tsx (hunk 10 above)
Comment (gtk-grafana) [reply public/app/features/explore/Logs/LogsTableActiveFields.tsx:63]: No, good catch, this is a stale comment.

File: public/app/features/explore/Logs/LogsTableNavField.tsx (hunk 12 above)
Comment (gtk-grafana) [reply public/app/features/explore/Logs/LogsTableNavField.tsx:31]: No this is another stale comment, I had previously added the ability to scroll oversized values in the initial implementations but it's not a pattern used commonly in Grafana so it was removed.

File: public/app/features/explore/Logs/LogsTableWrap.tsx (hunk 14 above)
Comment (gtk-grafana) [reply public/app/features/explore/Logs/LogsTableWrap.tsx:377]: Yeah I was looking at the wrong code 🤦  https://github.com/grafana/grafana/pull/79536/commits/d8866d276d08b80aa823cd5515f48c0297197790

Sorry about that. Fixed and pushed!

**Review by svennergr — APPROVED** [review]
Awesome! Thanks for the prompt changes 🚀🚀

**Pull request merged**